File: SDKs/defold/intelliversex/ivx_game_modes.lua
-- Copyright (c) 2026 Intelli-verse-X
-- MIT License — see LICENSE in the project root.

--- IntelliVerseX Game Modes for Defold
--- Lobby management, matchmaking, and player slots.
--- @module ivx_game_modes

local M = {}

-- ── Mode constants ──────────────────────────────────────────────────────────

M.SOLO              = 1
M.LOCAL_MULTIPLAYER = 2
M.ONLINE_VERSUS     = 3
M.ONLINE_COOP       = 4
M.RANKED            = 5
M.TURN_BASED        = 6

local MODE_NAMES = {
    [M.SOLO]              = "SOLO",
    [M.LOCAL_MULTIPLAYER] = "LOCAL_MULTIPLAYER",
    [M.ONLINE_VERSUS]     = "ONLINE_VERSUS",
    [M.ONLINE_COOP]       = "ONLINE_COOP",
    [M.RANKED]            = "RANKED",
    [M.TURN_BASED]        = "TURN_BASED",
}

-- ── State ───────────────────────────────────────────────────────────────────

local state = {
    current_mode    = M.SOLO,
    max_players     = 4,
    players         = {},
    match_active    = false,
    room_id         = nil,
    is_searching    = false,
}

local callbacks = {}


--- Register a callback: "mode_changed", "player_added", "player_removed",
--- "player_ready_changed", "match_started", "match_ended",
--- "match_found", "room_created", "room_joined", "room_left",
--- "room_list_updated", "search_cancelled"
--- @param event string
--- @param fn function
function M.on(event, fn)
    callbacks[event] = fn
end


--- Select a game mode.
--- @param mode number One of M.SOLO, M.LOCAL_MULTIPLAYER, etc.
--- @param max_players number? Default 4
function M.select_mode(mode, max_players)
    state.current_mode = mode
    state.max_players = max_players or 4
    state.players = {}
    state.match_active = false
    print("[IVXGameModes] Mode set to " .. (MODE_NAMES[mode] or "UNKNOWN") .. " (max " .. state.max_players .. " players)")
    _fire("mode_changed", mode)
end


--- Add a player to the next available slot.
--- @param display_name string
--- @param is_local boolean? Default true
--- @return table|nil Player info or nil if full
function M.add_player(display_name, is_local)
    if is_local == nil then is_local = true end
    if #state.players >= state.max_players then
        print("[IVXGameModes] Cannot add player — lobby full")
        return nil
    end
    local slot = #state.players + 1
    local info = {
        slot         = slot,
        display_name = display_name,
        is_local     = is_local,
        is_ready     = false,
    }
    table.insert(state.players, info)
    _fire("player_added", slot, info)
    return info
end


--- Remove a player by slot index.
--- @param slot number 1-based slot index
function M.remove_player(slot)
    if slot < 1 or slot > #state.players then
        print("[IVXGameModes] Invalid slot: " .. tostring(slot))
        return
    end
    table.remove(state.players, slot)
    for i = slot, #state.players do
        state.players[i].slot = i
    end
    _fire("player_removed", slot)
end


--- Set a player's ready state.
--- @param slot number 1-based slot index
--- @param ready boolean
function M.set_player_ready(slot, ready)
    if slot < 1 or slot > #state.players then
        print("[IVXGameModes] Invalid slot: " .. tostring(slot))
        return
    end
    state.players[slot].is_ready = ready
    _fire("player_ready_changed", slot, ready)
end


--- Start the match if all players are ready.
function M.start_match()
    if state.match_active then
        print("[IVXGameModes] Match already active")
        return
    end
    for _, p in ipairs(state.players) do
        if not p.is_ready then
            print("[IVXGameModes] Not all players are ready")
            return
        end
    end
    state.match_active = true
    print("[IVXGameModes] Match started with " .. #state.players .. " players")
    _fire("match_started")
end


--- End the current match.
function M.end_match()
    if not state.match_active then return end
    state.match_active = false
    print("[IVXGameModes] Match ended")
    _fire("match_ended")
end


--- Reset all state to defaults.
function M.reset()
    state.current_mode = M.SOLO
    state.max_players = 4
    state.players = {}
    state.match_active = false
    state.room_id = nil
    state.is_searching = false
end


-- ── Lobby ───────────────────────────────────────────────────────────────────

--- Create a new room/lobby.
--- @param name string
--- @param room_config table?
--- @param callback function?
function M.create_room(name, room_config, callback)
    local room_id = name:lower():gsub("%s", "_") .. "_" .. tostring(socket.gettime()):gsub("%.", "")
    state.room_id = room_id
    print("[IVXGameModes] Room created: " .. room_id)
    _fire("room_created", room_id)
    if callback then callback(room_id) end
end


--- Join an existing room.
--- @param room_id string
--- @param callback function?
function M.join_room(room_id, callback)
    state.room_id = room_id
    print("[IVXGameModes] Joined room: " .. room_id)
    _fire("room_joined", room_id)
    if callback then callback(room_id) end
end


--- List available rooms.
--- @param callback function(rooms)
function M.list_rooms(callback)
    local rooms = {}
    _fire("room_list_updated", rooms)
    if callback then callback(rooms) end
end


--- Leave the current room.
function M.leave_room()
    if not state.room_id then return end
    print("[IVXGameModes] Left room: " .. state.room_id)
    state.room_id = nil
    _fire("room_left")
end


-- ── Matchmaking ─────────────────────────────────────────────────────────────

--- Start searching for a match.
--- @param match_config table?
--- @param callback function?
function M.find_match(match_config, callback)
    if state.is_searching then
        print("[IVXGameModes] Already searching for a match")
        return
    end
    state.is_searching = true
    print("[IVXGameModes] Searching for match")
end


--- Cancel the current matchmaking search.
function M.cancel_search()
    if not state.is_searching then return end
    state.is_searching = false
    print("[IVXGameModes] Search cancelled")
    _fire("search_cancelled")
end


-- ── Getters ─────────────────────────────────────────────────────────────────

function M.get_current_mode() return state.current_mode end
function M.get_max_players() return state.max_players end
function M.get_players() return state.players end
function M.is_match_active() return state.match_active end
function M.get_room_id() return state.room_id end


-- ── Private ─────────────────────────────────────────────────────────────────

function _fire(event, ...)
    if callbacks[event] then
        callbacks[event](...)
    end
end


return M

File: SDKs/flutter/lib/src/ivx_game_modes.dart
import 'dart:async';

import 'types.dart';

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

/// Available game modes.
enum IVXGameMode {
  /// Classic single-player quiz.
  solo,

  /// Local couch co-op (shared device).
  localMultiplayer,

  /// Online real-time multiplayer.
  onlineMultiplayer,

  /// Cooperative team play.
  coop,

  /// Tournament bracket mode.
  tournament,

  /// Practice / training.
  practice,
}

// ---------------------------------------------------------------------------
// Models
// ---------------------------------------------------------------------------

/// A slot occupied (or reserved) by a player in the current match.
class IVXPlayerSlot {
  final int index;
  final String name;
  final bool isLocal;
  bool isReady;

  IVXPlayerSlot({
    required this.index,
    required this.name,
    this.isLocal = true,
    this.isReady = false,
  });

  @override
  String toString() =>
      'IVXPlayerSlot(index: $index, name: $name, ready: $isReady)';
}

/// Configuration snapshot for a match about to start.
class IVXMatchConfig {
  final IVXGameMode mode;
  final int maxPlayers;
  final List<IVXPlayerSlot> players;

  const IVXMatchConfig({
    required this.mode,
    required this.maxPlayers,
    required this.players,
  });

  @override
  String toString() =>
      'IVXMatchConfig(mode: ${mode.name}, players: ${players.length}/$maxPlayers)';
}

/// Describes a joinable lobby room.
class IVXRoomInfo {
  final String roomId;
  final String name;
  final int playerCount;
  final int maxPlayers;
  final IVXGameMode mode;
  final Map<String, dynamic> metadata;

  const IVXRoomInfo({
    required this.roomId,
    required this.name,
    required this.playerCount,
    required this.maxPlayers,
    required this.mode,
    this.metadata = const {},
  });

  @override
  String toString() =>
      'IVXRoomInfo(roomId: $roomId, name: $name, $playerCount/$maxPlayers)';
}

/// Summary produced when a match concludes.
class IVXMatchResult {
  final String matchId;
  final IVXGameMode mode;
  final List<IVXPlayerSlot> players;
  final Map<String, dynamic> scores;
  final Duration duration;

  const IVXMatchResult({
    required this.matchId,
    required this.mode,
    required this.players,
    this.scores = const {},
    this.duration = Duration.zero,
  });

  @override
  String toString() =>
      'IVXMatchResult(matchId: $matchId, mode: ${mode.name})';
}

// ---------------------------------------------------------------------------
// Manager
// ---------------------------------------------------------------------------

/// Manages game-mode selection, player slots, lobby rooms, and matchmaking.
///
/// ```dart
/// final modes = IVXGameModeManager.instance;
/// modes.selectMode(IVXGameMode.onlineMultiplayer, maxPlayers: 4);
/// modes.addPlayer('Alice');
/// modes.addPlayer('Bob', isLocal: false);
/// if (modes.canStartMatch) modes.startMatch();
/// ```
class IVXGameModeManager {
  static IVXGameModeManager? _instance;

  IVXGameMode _currentMode = IVXGameMode.solo;
  int _maxPlayers = 4;
  final List<IVXPlayerSlot> _players = [];
  bool _matchActive = false;
  String? _currentRoomId;
  String? _activeMatchId;
  int _matchCounter = 0;

  final StreamController<IVXGameMode> _modeController =
      StreamController<IVXGameMode>.broadcast();
  final StreamController<List<IVXPlayerSlot>> _playerController =
      StreamController<List<IVXPlayerSlot>>.broadcast();
  final StreamController<bool> _matchStateController =
      StreamController<bool>.broadcast();
  final StreamController<IVXRoomInfo> _roomController =
      StreamController<IVXRoomInfo>.broadcast();

  IVXGameModeManager._();

  /// Singleton accessor.
  static IVXGameModeManager get instance =>
      _instance ??= IVXGameModeManager._();

  /// Reset the singleton (useful for testing).
  static void resetInstance() {
    _instance?._dispose();
    _instance = null;
  }

  // ---------------------------------------------------------------------------
  // Streams
  // ---------------------------------------------------------------------------

  /// Fires whenever the active game mode changes.
  Stream<IVXGameMode> get onModeChanged => _modeController.stream;

  /// Fires whenever the player list changes.
  Stream<List<IVXPlayerSlot>> get onPlayersChanged => _playerController.stream;

  /// Fires `true` when a match starts, `false` when it ends.
  Stream<bool> get onMatchStateChanged => _matchStateController.stream;

  /// Fires when a room is created or joined.
  Stream<IVXRoomInfo> get onRoomEvent => _roomController.stream;

  // ---------------------------------------------------------------------------
  // State accessors
  // ---------------------------------------------------------------------------

  /// The currently selected game mode.
  IVXGameMode get currentMode => _currentMode;

  /// Unmodifiable view of current player slots.
  List<IVXPlayerSlot> get players => List.unmodifiable(_players);

  /// Whether a match is in progress.
  bool get isMatchActive => _matchActive;

  /// The maximum number of players for the current configuration.
  int get maxPlayers => _maxPlayers;

  /// `true` when at least one player is present and all are ready.
  bool get canStartMatch =>
      !_matchActive &&
      _players.isNotEmpty &&
      _players.every((p) => p.isReady);

  // ---------------------------------------------------------------------------
  // Mode selection
  // ---------------------------------------------------------------------------

  /// Choose a [mode] and optionally override [maxPlayers].
  void selectMode(IVXGameMode mode, {int maxPlayers = 4}) {
    if (_matchActive) {
      throw const IVXError(
        code: -1,
        message: 'Cannot change mode while a match is active.',
      );
    }
    _currentMode = mode;
    _maxPlayers = maxPlayers;
    _log('Mode set to ${mode.name} (max $maxPlayers)');
    _modeController.add(mode);
  }

  // ---------------------------------------------------------------------------
  // Player management
  // ---------------------------------------------------------------------------

  /// Add a player by [name] and return the assigned slot.
  IVXPlayerSlot addPlayer(String name, {bool isLocal = true}) {
    if (_players.length >= _maxPlayers) {
      throw IVXError(
        code: -1,
        message: 'Lobby full ($maxPlayers max).',
      );
    }
    final slot = IVXPlayerSlot(
      index: _players.length,
      name: name,
      isLocal: isLocal,
    );
    _players.add(slot);
    _log('Player added: $name (slot ${slot.index})');
    _playerController.add(players);
    return slot;
  }

  /// Remove the player at [slotIndex].
  void removePlayer(int slotIndex) {
    if (slotIndex < 0 || slotIndex >= _players.length) return;
    final removed = _players.removeAt(slotIndex);
    // Re-index remaining slots.
    for (var i = 0; i < _players.length; i++) {
      _players[i] = IVXPlayerSlot(
        index: i,
        name: _players[i].name,
        isLocal: _players[i].isLocal,
        isReady: _players[i].isReady,
      );
    }
    _log('Player removed: ${removed.name}');
    _playerController.add(players);
  }

  /// Mark [slot] as ready (or not).
  void setPlayerReady(int slot, bool ready) {
    if (slot < 0 || slot >= _players.length) return;
    _players[slot].isReady = ready;
    _playerController.add(players);
  }

  // ---------------------------------------------------------------------------
  // Match lifecycle
  // ---------------------------------------------------------------------------

  /// Start a match with the current players and mode.
  ///
  /// Throws if [canStartMatch] is `false`.
  void startMatch() {
    if (!canStartMatch) {
      throw const IVXError(
        code: -1,
        message: 'Cannot start: ensure at least one ready player and no '
            'active match.',
      );
    }
    _matchCounter++;
    _activeMatchId = 'match_${_matchCounter}_'
        '${DateTime.now().millisecondsSinceEpoch}';
    _matchActive = true;
    _log('Match started: $_activeMatchId (${_currentMode.name})');
    _matchStateController.add(true);
  }

  /// End the current match and return a [IVXMatchResult] summary.
  IVXMatchResult endMatch() {
    if (!_matchActive) {
      throw const IVXError(code: -1, message: 'No active match to end.');
    }
    final result = IVXMatchResult(
      matchId: _activeMatchId ?? 'unknown',
      mode: _currentMode,
      players: List.of(_players),
    );
    _matchActive = false;
    _activeMatchId = null;
    _log('Match ended');
    _matchStateController.add(false);
    return result;
  }

  /// Reset all state — mode, players, match — to defaults.
  void reset() {
    _matchActive = false;
    _activeMatchId = null;
    _currentRoomId = null;
    _players.clear();
    _currentMode = IVXGameMode.solo;
    _maxPlayers = 4;
    _log('Game mode manager reset');
    _modeController.add(_currentMode);
    _playerController.add(players);
    _matchStateController.add(false);
  }

  // ---------------------------------------------------------------------------
  // Lobby
  // ---------------------------------------------------------------------------

  /// Create a new lobby room with a display [name].
  Future<IVXRoomInfo> createRoom(String name, {IVXGameMode? mode}) async {
    final roomMode = mode ?? _currentMode;
    final room = IVXRoomInfo(
      roomId: 'room_${DateTime.now().millisecondsSinceEpoch}',
      name: name,
      playerCount: 1,
      maxPlayers: _maxPlayers,
      mode: roomMode,
    );
    _currentRoomId = room.roomId;
    _log('Room created: ${room.roomId}');
    _roomController.add(room);
    return room;
  }

  /// Join an existing room by [roomId].
  Future<IVXRoomInfo> joinRoom(String roomId) async {
    _currentRoomId = roomId;
    final room = IVXRoomInfo(
      roomId: roomId,
      name: roomId,
      playerCount: 0,
      maxPlayers: _maxPlayers,
      mode: _currentMode,
    );
    _log('Joined room: $roomId');
    _roomController.add(room);
    return room;
  }

  /// List available rooms.
  ///
  /// In a real implementation this would query a matchmaker service; the stub
  /// returns an empty list so consumers can code against the API now.
  Future<List<IVXRoomInfo>> listRooms() async => const [];

  /// Leave the currently-joined room.
  Future<void> leaveRoom() async {
    if (_currentRoomId != null) {
      _log('Left room: $_currentRoomId');
      _currentRoomId = null;
    }
  }

  // ---------------------------------------------------------------------------
  // Matchmaking
  // ---------------------------------------------------------------------------

  /// Request the matchmaker to find an appropriate match.
  ///
  /// Returns a [IVXRoomInfo] when a match is found.
  Future<IVXRoomInfo> findMatch({IVXGameMode? mode}) async {
    final searchMode = mode ?? _currentMode;
    _log('Searching for ${searchMode.name} match…');
    return IVXRoomInfo(
      roomId: 'mm_${DateTime.now().millisecondsSinceEpoch}',
      name: 'Matchmade Room',
      playerCount: 1,
      maxPlayers: _maxPlayers,
      mode: searchMode,
    );
  }

  /// Cancel an in-progress matchmaking search.
  Future<void> cancelSearch() async {
    _log('Matchmaking search cancelled');
  }

  // ---------------------------------------------------------------------------
  // Internal
  // ---------------------------------------------------------------------------

  void _dispose() {
    _modeController.close();
    _playerController.close();
    _matchStateController.close();
    _roomController.close();
  }

  void _log(String message) {
    // ignore: avoid_print
    print('[IntelliVerseX:GameModes] $message');
  }
}
